lib/nerve/service_watcher.rb

```diff
@@ -57,32 +59,59 @@ def initialize(service={})
       # force an initial report on startup
       @was_up = nil
 
+      # when this watcher is started it will store the
+      # thread here
+      @run_thread = nil
+      @should_finish = false
+
       log.debug "nerve: created service watcher for #{@name} with #{@service_checks.size} checks"
     end
 
+    def start()
+      unless @run_thread
+        @run_thread = Thread.new { self.run() }
+      else
+        log.error "nerve: tried to double start a watcher"
+      end
+    end
+
+    def stop()
+      log.info "nerve: stopping service watch #{@name}"
+      @should_finish = true
+      return true if @run_thread.nil?
+
+      unclean_shutdown = @run_thread.join(10).nil?
+      if unclean_shutdown
+        log.error "nerve: unclean shutdown of #{@name}, killing thread"
+        Thread.kill(@run_thread)
+      end
+      @run_thread = nil
+      !unclean_shutdown
+    end
+
+    def alive?()
+      !@run_thread.nil? && @run_thread.alive?
+    end
+
     def run()
       log.info "nerve: starting service watch #{@name}"
-
       @reporter.start()
 
-      until $EXIT
+      until $EXIT or @should_finish
```

Comment:
Done. Yea I know about the or thing I just sometimes get my python and ruby confused lol. Cheers.